cirq-core/cirq/ops/control_values.py

```diff
@@ -38,104 +39,99 @@ class AbstractControlValues(abc.ABC):
     """
 
     @abc.abstractmethod
-    def __and__(self, other: 'AbstractControlValues') -> 'AbstractControlValues':
-        """Sets self to be the cartesian product of all combinations in self x other.
-
-        Args:
-          other: An object that implements AbstractControlValues.
-
-        Returns:
-          An object that represents the cartesian product of the two inputs.
-        """
+    def validate(self, qid_shapes: Sequence[int]) -> None:
+        """Validates that all control values for ith qubit are in range [0, qid_shaped[i])"""
 
     @abc.abstractmethod
-    def _expand(self) -> Iterator[Tuple[int, ...]]:
-        """Expands the (possibly compressed) internal representation into a sum of products representation."""  # pylint: disable=line-too-long
+    def expand(self) -> 'SumOfProducts':
+        """Returns an expanded `cirq.SumOfProduct` representation of this control values."""
 
+    @property
     @abc.abstractmethod
-    def diagram_repr(self, label: Optional[str] = None) -> str:
-        """Returns a string representation to be used in circuit diagrams."""
+    def is_trivial(self) -> bool:
+        """Returns True iff each controlled variable is activated only for value 1.
 
-    @abc.abstractmethod
-    def _number_variables(self) -> int:
-        """Returns the number of variables controlled by the object."""
+        This configuration is equivalent to `cirq.SumOfProducts(((1,) * num_controls))`
+        and `cirq.ProductOfSums(((1,),) * num_controls)`
 
-    @abc.abstractmethod
-    def __len__(self) -> int:
-        pass
+        """
 
     @abc.abstractmethod
-    def _identifier(self) -> Any:
-        """Returns the internal representation of the object."""
+    def _num_qubits_(self) -> int:
+        """Returns the number of qubits for which control values are stored by this object."""
 
     @abc.abstractmethod
-    def __hash__(self) -> int:
-        pass
+    def _json_dict_(self) -> Dict[str, Any]:
+        """Returns a dictionary used for serializing this object."""
 
     @abc.abstractmethod
-    def __repr__(self) -> str:
-        pass
+    def _circuit_diagram_info_(
+        self, args: 'cirq.CircuitDiagramInfoArgs'
+    ) -> 'cirq.CircuitDiagramInfo':
+        """Returns information used to draw this object in circuit diagrams."""
 
     @abc.abstractmethod
-    def validate(self, qid_shapes: Union[Tuple[int, ...], List[int]]) -> None:
-        """Validates control values
-
-        Validate that control values are in the half closed interval
-        [0, qid_shapes) for each qubit.
+    def __iter__(self) -> Iterator[Tuple[int, ...]]:
+        """Iterator on internal representation of control values used by the derived classes.
```

Comment:
This seems dangerous for `__iter__` to yield the internal representation, since special dunder methods are part of the public interface of a class. Previously, this called `self._expand` so it would always yield tuples from the `SumOfProducts` representation, which I think made more sense. If users need access to the internal representation, I think they should directly iterate over the internal fields.

Reply:
I agree, but the problem with returning the expanded representation is that it will break all user code of the form 

```
>>> for cv in controlled_gate.control_values:
            # cv right now is a iterating on product of sums but will later iterate on sum of products. 
```

This will also be a silent break (no deprecation cycle; no warnings; no API breakages).

Reply:
We, any code that is currently iterating expects to get values from `_expand()` (e.g. from the `SumOfProducts` representation). If we keep `__iter__` I think we should stick with that rather than changing to iterate over the internal representation directly.